File: src/main/java/com/cartrecipes/model/Recipe.java
package com.cartrecipes.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;

import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "recipes")
@Getter
@Setter
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Recipe {
    @Id
    @EqualsAndHashCode.Include
    private Long id;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "price_in_cents", nullable = false)
    @Min(0)
    private Integer priceInCents;

    @ManyToMany
    @JoinTable(
            name = "recipe_items",
            joinColumns = @JoinColumn(name = "recipe_id", referencedColumnName = "id", nullable = false),
            inverseJoinColumns = @JoinColumn(name = "product_id", referencedColumnName = "id", nullable = false),
            uniqueConstraints = @UniqueConstraint(columnNames = {"recipe_id", "product_id"}, name = "uk_recipe_items")
    )
    private Set<Product> products = new HashSet<>();
}

File: src/main/kotlin/com/cartrecipes/model/Recipe.kt
package com.cartrecipes.model

import jakarta.persistence.*
import jakarta.validation.constraints.Min

@Entity
@Table(name = "recipes")
class Recipe (
    @field:Id
    val id: Long,

    @field:Column(name = "name", nullable = false)
    val name: String,

    @field:Column(name = "price_in_cents", nullable = false)
    @field:Min(0)
    var priceInCents: Int,

    @field:ManyToMany
    @field:JoinTable(
        name = "recipe_items",
        joinColumns = [JoinColumn(name = "recipe_id", referencedColumnName = "id", nullable = false)],
        inverseJoinColumns = [JoinColumn(name = "product_id", referencedColumnName = "id", nullable = false)],
        uniqueConstraints = [UniqueConstraint(columnNames = ["recipe_id", "product_id"], name = "uk_recipe_items")]
    )
    val products: MutableSet<Product> = mutableSetOf()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Recipe

        return id == other.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }
}